src/core/unit-printf.c

```diff
@@ -299,3 +299,37 @@ int unit_full_printf(const Unit *u, const char *format, char **ret) {
 
         return specifier_printf(format, table, u, ret);
 }
+
+int unit_default_printf(const char *format, char **ret) {
+        /* This is similar to unit_full_printf(), but is used for setting default environments outside of any
+         * invididual unit. As such, only specifiers that are not specific to a unit are supported.
+         */
+
+        assert(format);
+        assert(ret);
+
+        const Specifier table[] = {
+                { 'g', specifier_group_name,               NULL },
+                { 'G', specifier_group_id,                 NULL },
+                { 'U', specifier_user_id,                  NULL },
+                { 'u', specifier_user_name,                NULL },
+                { 'h', specifier_user_home,                NULL },
+                { 's', specifier_user_shell,               NULL },
+
+                { 't', specifier_special_directory,        UINT_TO_PTR(EXEC_DIRECTORY_RUNTIME) },
+                { 'S', specifier_special_directory,        UINT_TO_PTR(EXEC_DIRECTORY_STATE) },
+                { 'C', specifier_special_directory,        UINT_TO_PTR(EXEC_DIRECTORY_CACHE) },
+                { 'L', specifier_special_directory,        UINT_TO_PTR(EXEC_DIRECTORY_LOGS) },
+                { 'E', specifier_special_directory,        UINT_TO_PTR(EXEC_DIRECTORY_CONFIGURATION) },
+                { 'T', specifier_tmp_dir,                  NULL },
+                { 'V', specifier_var_tmp_dir,              NULL },
+
+                { 'm', specifier_machine_id,               NULL },
+                { 'H', specifier_host_name,                NULL },
+                { 'b', specifier_boot_id,                  NULL },
+                { 'v', specifier_kernel_release,           NULL },
+                {}
```

Comment:
hmm, another option would be to beef up all specifier_xyz() functions that expect a unit pointer to handle a unit pointer of NULL properly, and generate a warning then. that would allow having the same table in all cases.